lib/pure/os.nim

```diff
@@ -234,45 +235,79 @@ proc findExe*(exe: string, followSymlinks: bool = true;
-              if len < 0:
-                raiseOSError(osLastError(), exe)
-              if len > maxSymlinkLen:
-                r = newString(len+1)
-                len = readlink(x.cstring, r.cstring, len)
-              setLen(r, len)
-              if isAbsolute(r):
-                x = r
+      checkCurrentDir()
+    let path = getEnv("PATH")
+    for candidate in split(path, PathSep):
+      if candidate.len == 0: continue
+      when defined(windows):
+        var x = (if candidate[0] == '"' and candidate[^1] == '"':
+                  substr(candidate, 1, candidate.len-2) else: candidate) /
+                exe
+      else:
+        var x = expandTilde(candidate) / exe
+      for ext in extensions:
+        var x = addFileExt(x, ext)
+        if fileExists(x):
+          when not (defined(windows) or defined(nintendoswitch)):
+            while followSymlinks: # doubles as if here
+              if x.symlinkExists:
+                var r = newString(maxSymlinkLen)
+                var len = readlink(x.cstring, r.cstring, maxSymlinkLen)
+                if len < 0:
+                  raiseOSError(osLastError(), exe)
+                if len > maxSymlinkLen:
+                  r = newString(len+1)
+                  len = readlink(x.cstring, r.cstring, len)
+                setLen(r, len)
+                if isAbsolute(r):
+                  x = r
+                else:
+                  x = parentDir(x) / r
               else:
-                x = parentDir(x) / r
-            else:
-              break
-        return x
-  result = ""
+                break
+          yield x
+
+proc findExe*(exe: string, followSymlinks: bool = true;
+        extensions: seq[string]=ExeExtsSeq): string {.
+  tags: [ReadDirEffect, ReadEnvEffect, ReadIOEffect], noNimJs, raises: [OSError].} =
+  ## Searches for `exe` in the current working directory and then
+  ## in directories listed in the ``PATH`` environment variable.
+  ##
+  ## Returns `""` if the `exe` cannot be found. `exe`
+  ## is added the `ExeExts <#ExeExts>`_ file extensions if it has none.
+  ##
+  ## If the system supports symlinks it also resolves them until it
+  ## meets the actual file. This behavior can be disabled if desired
+  ## by setting `followSymlinks = false`.
+  return iterExes(exe, followSymlinks, extensions)()
+
+proc findExeAll*(exe: string, followSymlinks: bool = true;
+        extensions: seq[string]=ExeExtsSeq): seq[string] {.
+  tags: [ReadDirEffect, ReadEnvEffect, ReadIOEffect], noNimJs.} =
+  ## Searches for `exe` in the current working directory and then
+  ## in directories listed in the ``PATH`` environment variable.
+  ##
+  ## Returns `""` if the `exe` cannot be found. `exe`
+  ## is added the `ExeExts <#ExeExts>`_ file extensions if it has none.
+  ##
+  ## If the system supports symlinks it also resolves them until it
+  ## meets the actual file. This behavior can be disabled if desired
+  ## by setting `followSymlinks = false`.
```

Comment:
This needs some kind of max limit integer Positive for safety.

I can just `touch` and `chmod +x` a ton of files on a folder,
and the function will hang the system for years...